### PR TITLE
Annotation Idempotency

### PR DESCRIPTION
**Pull request by jihan-yin**
Adds `annotation_id` as optional parameter to annotations and prediction classes. This will be used to uniquely identify each annotation/prediction within its dataset item. Also adds `update` as an optional boolean parameter when uploading predictions and annotations. Setting `update=False`, the default behavior, will skip uploading annotations to dataset images that already have annotations with the same `annotation_id`. Setting `update=True` will overwrite such annotations.

### CODE REVIEW

File: nucleus/__init__.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def __init__(self, api_key: str, use_notebook: bool = False):
         self.api_key = api_key
         self.tqdm_bar = tqdm.tqdm
         if use_notebook:
-            self.tqdm_bar = tqdm.notebook.tqdm
+            self.tqdm_bar = tqdm_notebook.tqdm
```
Comment (jihan-yin) [inline nucleus/__init__.py:131]: Was running into issues calling the api from within a notebook, this resolved it for me

File: nucleus/__init__.py [hunk 2]
```diff
@@ -462,13 +462,15 @@ def exception_handler(request, exception):
     def annotate_dataset(
         self,
         dataset_id: str,
-        annotations: List[DatasetItem],
+        annotations: List[Union[BoxAnnotation, PolygonAnnotation]],
```
Comment (sasha-scale) [inline nucleus/__init__.py:465]: omg - nice fix

File: nucleus/__init__.py [hunk 3]
```diff
@@ -462,13 +462,15 @@ def exception_handler(request, exception):
...
+        update: bool,
```
Comment (sasha-scale) [inline nucleus/__init__.py:466]: see comment on sister PR - I think this should be optional with default value False

File: nucleus/__init__.py [hunk 4]
```diff
@@ -602,7 +600,11 @@ def predict(
         tqdm_batches = self.tqdm_bar(batches)
 
         for batch in tqdm_batches:
-            batch_payload = {ANNOTATIONS_KEY: batch}
+            batch_payload = construct_box_predictions_payload(
+                annotations,
+                update,
+            )
+            print(batch_payload)
```
Comment (sasha-scale) [inline nucleus/__init__.py:607]: remove print statement before merging

File: nucleus/constants.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
-NUCLEUS_ENDPOINT = "https://api.scale.com/v1/nucleus"
+#NUCLEUS_ENDPOINT = "https://api.scale.com/v1/nucleus"
```
Comment (sasha-scale) [inline nucleus/constants.py:1]: remove before merging

File: nucleus/dataset.py [hunk 6]
```diff
@@ -10,11 +10,14 @@
     REFERENCE_IDS_KEY,
     NAME_KEY,
     ITEM_KEY,
+    DEFAULT_ANNOTATION_UPDATE_MODE,
     ANNOTATIONS_KEY,
 )
 from .payload_constructor import construct_model_run_creation_payload
 
 
+
```
Comment (sasha-scale) [inline nucleus/dataset.py:19]: :nit: extraneous white space

**Review by sasha-scale — CHANGES_REQUESTED** [review]
I was unable to upload GT annotations without specifying the annotation_id, the error message was "annotation_id must be a string."  I think the issue was that when I did not specify it, it got sent as `None` in the request payload which didn't make it through schema validation

**Review by sasha-scale — APPROVED** [review]
LGTM, tested locally.  Response format does not have annotations_ignored which to me blocks merge, but that is more a comment on the scaleapi Pr than this one.

The unit tests are incredibly thorough!  great jobs!

File: tests/helpers.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    assert annotation_instance.label == annotation_dict['label']
+    assert annotation_instance.annotation_id == annotation_dict['annotation_id']
+    for instance_pt, dict_pt in zip(annotation_instance.vertices, annotation_dict['vertices']):
+        assert instance_pt['x'] == dict_pt['x']
+        assert instance_pt['y'] == dict_pt['y']
+
+# Asserts that a box prediction instance matches a dict representing its properties.
+# Useful to check prediction uploads/updates match.
+def assert_box_prediction_matches_dict(prediction_instance, prediction_dict):
+    assert_box_annotation_matches_dict(prediction_instance, prediction_dict)
```
Comment (rkaplan) [inline tests/helpers.py:85]: Infinite recursion?

File: tests/helpers.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    assert prediction_instance.confidence == prediction_dict['confidence']
+
+def assert_polygon_prediction_matches_dict(prediction_instance, prediction_dict):
+    assert_polygon_annotation_matches_dict(prediction_instance, prediction_dict)
```
Comment (rkaplan) [inline tests/helpers.py:89]: Infinite recursion?

File: tests/helpers.py (hunk 7 above)
Comment (jihan-yin) [reply tests/helpers.py:85]: It's a different function 😄

File: tests/helpers.py (hunk 8 above)
Comment (jihan-yin) [reply tests/helpers.py:89]: same as above

**Review by rkaplan — APPROVED** [review]

**Pull request merged**